core-api/src/main/java/com/optimizely/ab/bucketing/DecisionService.java

```diff
@@ -136,42 +143,49 @@ public Variation getVariation(@Nonnull Experiment experiment,
-                variation = getStoredVariation(experiment, userProfile, projectConfig, reasons);
+                decisionVariation = getStoredVariation(experiment, userProfile, projectConfig);
+                reasons.merge(decisionVariation.getReasons());
+                variation = decisionVariation.getResult();
                 // return the stored variation if it exists
                 if (variation != null) {
-                    return variation;
+                    return new DecisionResponse(variation, reasons);
                 }
             } else { // if we could not find a user profile, make a new one
                 userProfile = new UserProfile(userId, new HashMap<String, Decision>());
             }
         }
 
-        if (ExperimentUtils.doesUserMeetAudienceConditions(projectConfig, experiment, filteredAttributes, EXPERIMENT, experiment.getKey(), reasons)) {
-            String bucketingId = getBucketingId(userId, filteredAttributes, reasons);
-            variation = bucketer.bucket(experiment, bucketingId, projectConfig, reasons);
+        DecisionResponse<Boolean> decisionMeetAudience = ExperimentUtils.doesUserMeetAudienceConditions(projectConfig, experiment, filteredAttributes, EXPERIMENT, experiment.getKey());
+        reasons.merge(decisionMeetAudience.getReasons());
+        if (decisionMeetAudience.getResult()) {
+            String bucketingId = getBucketingId(userId, filteredAttributes);
+
+            decisionVariation = bucketer.bucket(experiment, bucketingId, projectConfig);
+            reasons.merge(decisionVariation.getReasons());
+            variation = decisionVariation.getResult();
 
             if (variation != null) {
                 if (userProfileService != null && !ignoreUPS) {
-                    saveVariation(experiment, variation, userProfile, reasons);
+                    saveVariation(experiment, variation, userProfile);
                 } else {
                     logger.debug("This decision will not be saved since the UserProfileService is null.");
                 }
             }
 
-            return variation;
+            return new DecisionResponse(variation, reasons);
         }
 
         String message = reasons.addInfo("User \"%s\" does not meet conditions to be in experiment \"%s\".", userId, experiment.getKey());
         logger.info(message);
-        return null;
+        return new DecisionResponse(null, reasons);
     }
 
     @Nullable
-    public Variation getVariation(@Nonnull Experiment experiment,
-                                  @Nonnull String userId,
-                                  @Nonnull Map<String, ?> filteredAttributes,
-                                  @Nonnull ProjectConfig projectConfig) {
-        return getVariation(experiment, userId, filteredAttributes, projectConfig, Collections.emptyList(), DefaultDecisionReasons.newInstance());
+    public DecisionResponse<Variation>  getVariation(@Nonnull Experiment experiment,
```

Comment:
These methods are all marked as nullable return value.  However, the usage does not test for null:

```
Variation variation = decisionService.getVariation(experiment, userId, copiedAttributes, projectConfig).getResult();

Reply:
I'll change them all @Nonnull.